single.php

```diff
@@ -130,6 +130,35 @@
 
 if ( post_password_required( $post->ID ) ) {
 
+	// Check if page url has a unique id(custom hash), appended with it, if not add one.
+	$custom_hash = filter_input( INPUT_GET, 'ch', FILTER_SANITIZE_STRING );
+	if ( ! $custom_hash ) {
+		wp_safe_redirect( add_query_arg( 'ch', md5( uniqid( '', true ) ), get_permalink() ) );
```

Comment:
Why do we need to redirect here?

Reply:
I guess we can use the `$custom_hash` directly in the same request?

Reply:
When visitor submit the form with password, it goes to [wp-login.php](https://github.com/WordPress/WordPress/blob/master/wp-login.php#L803) and redirect back to same page. if validation fails, we track the request based on hash token from page url.